src/character.cpp

```diff
@@ -7573,13 +7590,13 @@ int Character::get_armor_fire( body_part bp ) const
 
 void Character::did_hit( Creature &target )
 {
-    enchantment_cache.cast_hit_you( *this, target.pos() );
+    enchantment_cache.cast_hit_you( *this, target );
 }
 
-void Character::on_hit( Creature * /*source*/, body_part /*bp_hit*/,
+void Character::on_hit( Creature *source, body_part /*bp_hit*/,
                         float /*difficulty*/, dealt_projectile_attack const *const /*proj*/ )
 {
-    enchantment_cache.cast_hit_me( *this );
+    enchantment_cache.cast_hit_me( *this, *source);
```

Comment:
You should check that `source` is actually non-null before dereferencing it.

Reply:
hmm, some of the spells ignores the source, and are cast on self, not sure how to proceed

Reply:
Doing `*source` causes undefined behaviour if `source` is a null pointer. You need at least something like `if( source == nullptr ) { return; }`.

Reply:
the think is i don't want to return, i would  gladly do that inside the cast_hit_me,
can i just check for nullptr here and pass null if its nullptr?
or do i have to create an overload with a single parameter for that?

Reply:
what i really want to know: will it still be undefined behaviour if i check `if(&source != nullptr)` inside the function?

Reply:
Yes, dereferencing a null pointer is undefined behaviour, so you've already triggered it by the time you go into the function.

Reply:
what are my options? i'm c# dev by trade so I'm not sure how do deal with this situation. Is an overload function the only way?

Reply:
you could pass a pointer as a parameter and then check if it's null inside, or you can write an overload function.

Reply:
All you need is this.
```suggestion
    if( source == nullptr ) {
        return;
    }
    enchantment_cache.cast_hit_me( *this, *source);
```

Reply:
I think i handled this by passing a pointer as @KorGgenT suggested by passing a pointer as a param